Feature/text skill implementation

AdaraPutri opened the pull request:
## 📝 Description

MASSIVE REFACTOR and new implementation on how skills are analyzed for text files. Following the structure of PR #212, this PR fully implements the text-skill detector system and cleans up the entire text analysis pipeline.

Key changes:

- Reworked the flow in project_analysis.py so that ALL text files are passed into text_analyze.py regardless of LLM consent.
  - → LLM consent now only affects the main file summary (generated vs user-prompted).
  - → Skill extraction is fully decoupled from summary generation.

- Added scripts (brand new):
  - text_analyze.py
  - llm_summary.py
  - alt_summary.py
  - text_detector_registry.py
  - text_detectors.py

- Modified scripts:
  - alt_analyze.py → now only handles lexical diversity + reading level
  - csv_analyze.py → no direct printing; new analyze_all_csv() returns aggregated CSV metadata
  - bucket_types.py → updated to support detector-based scoring
  - text_buckets.py → updated with 10 text buckets wired to the new detectors
  - skill_extraction.py → routed to new text skill flow
  - text_skill_extraction.py → now aggregates detector scores, evidence, and returns final results to text_analyze.py

All ten detectors (clarity, structure, vocabulary, argumentation, depth, iterative process, planning behavior, research, data collection, and data analysis) now use multi-criteria scoring with rich evidence, covering writing quality, organization, reasoning, research depth, and whether the student collected and analyzed data. → PLEASE CHECK documentation in the text_detectors.py for detailed info of how each of the score is calculated

NOTE:
- This pr does not use store_text_offline_metrics, store_text_llm_metrics as those will need to be updated first


TEST SCRIPT UPDATES:
I updated test_alt_analyze.py, test_csv_analyze.py, and test_text_analyze.py so they accurately match the new analysis flow and removed features. For test_alt_analyze.py, I removed tests for deprecated functions (topic_extraction, extract_keywords, project metrics, etc.) and rewrote the suite to only verify the current linguistic-complexity outputs. For test_csv_analyze.py, I updated all tests to reflect the simplified CSV metadata structure (no numeric stats, no printing), added correct row/column/missingness checks, and adapted growth-trend grouping behavior. For test_text_analyze.py, I rewrote the entire suite to test the new run_text_pipeline instead of the old LLM analysis pipeline by mocking CSV metadata, main-file selection, input(), and skill-extraction; ensuring summaries and bucket scores are returned; and fixing patch ordering issues so tests hit the correct branches. Tests now all pass.

**Closes:** #214 

---

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [x] ⚡ Performance improvement

---

## 🧪 Testing

- [x] Manual testing with text project files
- [x] python -m pytest passes

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots
Terminal should now ONLY print file summary AND skills (score is printed for debug purposes right now)
<img width="668" height="665" alt="Screenshot 2025-11-21 at 10 49 49 AM" src="https://github.com/user-attachments/assets/e461fad1-7fa7-4c78-82c8-5f33c13d35db" />

This is how the database should look like in project_skills table:
<img width="903" height="244" alt="Screenshot 2025-11-21 at 10 50 15 AM" src="https://github.com/user-attachments/assets/841c97e3-dd81-4e48-be4d-354225e5f59b" />

Tests passing:
<img width="948" height="771" alt="Screenshot 2025-11-23 at 6 25 08 AM" src="https://github.com/user-attachments/assets/69b75485-f372-44c5-8c09-a58412aaf068" />

AdaraPutri committed:
specified text skill detectors

AdaraPutri committed:
refactor csv_analyze, and separated summary generator scripts out of main analysis flow

AdaraPutri committed:
fixed new flow in project_analysis for text files

AdaraPutri committed:
clean up alt_analyze to only keep linguistic complexity analysis for skill extraction

AdaraPutri committed:
expanded text skill detectors and printed skills to terminal

AdaraPutri committed:
delete old test files

taoTimTim commented on src/analysis/skills/detectors/text/text_detectors.py:
```diff
@@ -0,0 +1,473 @@
+    return {"score": score / total, "evidence": [ev]}
+
+
+def detect_depth_of_content(file_text: str, file_name: str):
```
The depth detector currently relies on substring matching and an unstable idea-density calculation, which can creates false positives (e.g., “modeling” triggers “model”). I recommend using whole word boundaries and adding a minimum word count check so the score is less sensitive to short texts.

```
# safer whole-word match
if re.search(rf"\b{kw}\b", text_lower):
    abstract_count += 1

# stabilize idea density for short texts
if len(words) < 80:
    return {"score": 0, "evidence": [{"reason": "text too short"}]}
```

taoTimTim commented on src/analysis/skills/detectors/text/text_detectors.py:
```diff
@@ -0,0 +1,473 @@
+
+
+
+def detect_iterative_process(file_text: str, file_name: str, supporting_files=None):
```
This detector currently scores drafts based on filename substrings (“draft”, “rev”), which can produce misleading results. A file named “final_draft_v3.txt” will score higher than a user with multiple real drafts that don't contain the keywords. I recommend changing the heuristics or add a basic version-number detection:

```
# safer version detection
if re.search(r"(v\d+|version\s*\d+)", fname):
    versions.append(fname)

# restrict draft detection to whole words
if re.search(r"\bdraft\b|\brev(ision)?\b", fname):
    drafts.append(fname)
```

You could make this code even more robust too.

taoTimTim commented on src/analysis/skills/flows/skill_extraction.py:
```diff
@@ -0,0 +1,41 @@
+    if project_type == "code":
+        extract_code_skills(conn, user_id, project_name, classification, files)
+    elif project_type == "text":
+        extract_text_skills(conn, user_id, project_name, classification, files)
```
There is a difference in attribuets being passed through. Change this line to match what the function `extract_text_skills` actually wants. The function header is this:

```
def extract_text_skills(
    main_text,
    supporting_texts,
    csv_metadata,
    project_name,
    user_id,
    conn,
):
```

OR change the function except just the metadata and retrieve the data inside the function `extract_text_skills`

taoTimTim reviewed (changes requested):
This is a huge PR, thank you for working on this, Adara! I appreciate the full text pipeline refactoring and rework.

I am getting some test failures, I totally understand adding the tests in a new PR so I am unsure if you plan on fixing the failing ones in this PR or the old one. This is what I am getting:
<img width="1322" height="180" alt="image" src="https://github.com/user-attachments/assets/22b5acae-6039-4a97-bc11-456eb925fbda" />

I also noticed that none of the skills are being saved to the database, and I suspect it is because of an attribute mismatch in the code. Line 37 of skill_extraction.py sends attributes that the actual function `extract_text_skills` does not expect or does not match. Feel free to decide the best way to handle this, either changing the call in `skill_extraction.py` or changing the function.

I left a bunch of inline comments on the detectors with some ideas on how to improve them, obviously this is a ton of work so I tried to make it clear on the best way to implement the detectors would be. Not all of them have to be changed, but doing so would make our system better at detecting skills overall.

This was a TON of work, and the refactoring you did is well done. I would definitely like to see this rebased with main, to better see what you changed. As discussed on Discord you plan to do this as soon as you get the chance, which is great, and I can maybe see the changes you made a bit better when I review this again.

It made sense as to why you removed a bunch of tests, and putting the new tests in a new PR makes sense for PR size management, but do take a look at the failing tests.

This is a long review, but it was a big PR. Thanks again for working on this, I hope it was not too difficult to understand what the skill pipeline is doing.

AdaraPutri committed:
Merge remote-tracking branch 'origin/main' into feature/text-skill-implementation

AdaraPutri committed:
modified new text pipeline to return what project summary needs

AdaraPutri committed:
updated project analysis to store project summary better

AdaraPutri committed:
Merge remote-tracking branch 'origin/main' into feature/text-skill-implementation

AdaraPutri committed:
implemented all detector function improvement suggestions

AdaraPutri commented:
@taoTimTim hi Timmi, thank you so much for the thorough feedback! I have implemented all the suggestions to make the detector functions more robust and accurate. Regarding the db storing, the skills are currently already being saved in the database (I've provided a screenshot of project_skills table in the pr description and it is updated accordingly). 

The pipeline I've implemented does not call `skill_extraction.py` in `project_analysis.py` (the `_run_skill_extraction_for_all()` function call is commented out because it analyzes the skills for all projects at the very end, printing the results not with each project summary). So I've changed the flow so that `text_skill_extraction.py` is called within `text_analyze.py` (this let's us see the skill detection alongside each project in the same terminal printing section). Let me know if you think I should keep this approach or not!

I will be putting up another pr to fix the failing tests 🙌

<img width="827" height="120" alt="Screenshot 2025-11-23 at 5 39 24 AM" src="https://github.com/user-attachments/assets/0f49be14-3a51-4692-bacc-d11271bb9df4" />

AdaraPutri committed:
modified test script for updated csv_analyze

AdaraPutri committed:
restored test_alt_analyze.py, test_text_llm_analyze.py (renamed as test_text_analyze.py) and updated test cases in them to properly test the updated functions

taoTimTim reviewed (approved):
Amazing changes here, Adara. I really appreciate the work you put in.

The changes you made to the detectors are great, they're a bit more detailed than what was originally implemented which is nice.

All the tests are passing, it was a good call to fix those here due to the PR's waiting for review. Thank you for adding the detailed write-up about them inside the description, that was super helpful!

After conversing with you over Discord over the function attribute mismatch, you informed me it was unrelated to your PR as of this moment, and should be fixed in a separate PR. I completely agree, especially due to the size of this PR already.

I have no other comments, this is some great work. Approved, nice work!

ammaarkhan reviewed (changes requested):
Hey Adara, this is really impressive work! The four-stage text analysis pipeline in text_analyze.py is super clean, and the detector registry pattern makes it easy to add or remove detectors in the future.

The 10 detectors you've built out are comprehensive, covering writing mechanics to research rigor with clear scoring rubrics. 

A couple things I noticed:

Debug CSV printing: There's some CSV metadata printing left in `text_analyze.py:57-65` that looks like debug code. Should we remove that before merging?

Testing failure: Running pytest currently fails with a recursion error caused by how test_analyze_all_csv mocks os.path.join at line 106. The mock returns a pathlib Path object instead of a string, which creates infinite recursion when pathlib tries to process it internally. Not sure if I'm the only one facing this? 
A fix I found is to mock os.path.dirname and os.path.abspath instead (the specific functions that analyze_all_csv uses), and create the proper zip_data/test_project/ directory structure that the function expects. I've attached the fixed function at the bottom of the review for reference. 

Error on running pytest:
<img width="950" height="356" alt="Image" src="https://github.com/user-attachments/assets/dd290ce6-97bb-4abb-bcee-d2c42b1b0481" />

This is a really strong foundation for the skill extraction system. Once we clean up that debug printing and testing issue, this is ready to go!

Suggested solution reference:

```python
def test_analyze_all_csv(tmp_path, monkeypatch):
    # Create a proper directory structure that matches what analyze_all_csv expects
    # It expects: ZIP_DATA_DIR / zip_name / file_path
    zip_data_dir = tmp_path / "zip_data"
    zip_data_dir.mkdir()

    project_dir = zip_data_dir / "test_project"
    project_dir.mkdir()

    # Fake parsed_files
    parsed_files = [
        {"file_name": "data1.csv", "file_path": "data1.csv"},
        {"file_name": "data2.csv", "file_path": "data2.csv"},
    ]

    df1 = pd.DataFrame({"A": [1, 2]})
    df2 = pd.DataFrame({"A": [1, 2, 3]})

    df1.to_csv(project_dir / "data1.csv", index=False)
    df2.to_csv(project_dir / "data2.csv", index=False)

    # Mock only the directory paths that analyze_all_csv computes
    def mock_dirname(_path):
        # Return fake repo root when asked for parent directories
        return str(tmp_path)

    def mock_abspath(path):
        return str(path)

    monkeypatch.setattr(ca.os.path, "dirname", mock_dirname)
    monkeypatch.setattr(ca.os.path, "abspath", mock_abspath)

    result = ca.analyze_all_csv(parsed_files, zip_path=str(tmp_path / "test_project.zip"))

    assert result["growth_trend_present"] is True
    assert "data" in result["growth_trends"]
    assert result["files"][0]["col_count"] == 1
```

AdaraPutri committed:
commented out csv metadata debugs and fixed test_analyze_all_csv script

ammaarkhan reviewed (approved):
Thanks for making the changes Adara. The tests are now passing. Ready to merge!

<img width="962" height="278" alt="Image" src="https://github.com/user-attachments/assets/bf823441-294e-4bc7-8846-f3cb83cb86ee" />
